### DISCUSSION TITLE
What is a recommended way of updating detekt config after a new release?

### DISCUSSION THREAD

**Discussion by petertrr** [detekt/detekt#3558] [Q&A] [answered]
I have a custom yml config in my project, and every time a new version of detekt is released, I can't figure a good way to update it. There are usually new rules, which are disabled by default. This is fine, because it unexpected checks won't be activated, but requires to manually add new entries into the configuration file.
So, in order to do so, I should either go through release notes, which don't have a separate section for new rules and usually have quite a long list of changes, or run config generation task, which would overwrite my custom config. So I wonder, if there is any recommended way of updating? Or a more convenient way to retrieve a list of new rules from ma particular release?

**Comment by BraisGabin** [comment, marked as answer]
Really good question! I want to read the answers of other people. What I do is to use the GNU `diff` and `patch` commands.

- `mv config/detekt/detekt.yml config/detekt/mydetekt.yml`
  - Move my configuration to another file
- `./gradlew detektGenerateConfig`
  - Recreate the default configuration
- `diff -u config/detekt/detekt.yml config/detekt/mydetekt.yml >config/detekt/mydetekt.patch`
  - Save the difference between the default and my configuration
- `rm config/detekt/detekt.yml config/detekt/mydetekt.yml`
  - remove all the configurations
- Update detekt
- `./gradlew detektGenerateConfig`
  - Recreate the default configuration with the new detekt version
- `patch -u --merge --no-backup-if-mismatch config/detekt/detekt.yml <config/detekt/mydetekt.patch`
  - Apply the old changes over the new file. Be careful here! You should use the gnu patch. The mac patch doesn't support the `--merge` option.
- `rm config/detekt/mydetekt.patch`
  - Remove the patch

I know that this is far from ideal so I'd like to know how other people do this and find better ways to improve this.

**Reply by petertrr** [reply]
Thanks for the reply! That's a nice approach, I'll definitely create a similar script for myself :)
The one downside is, though, that it loses advantage of yaml being structured text, I think this comparison can be more effectively done inside detekt itself. And I'll be intereseted to hear other replies as well.

**Comment by cortinico** [comment]
> So I wonder, if there is any recommended way of updating? Or a more convenient way to retrieve a list of new rules from ma particular release?

Have you considered using `buildUponDefaultConfig` to maintain only the subset of changes rather than the entire config file?
https://detekt.github.io/detekt/gradle.html#options-for-detekt-configuration-closure

**Reply by petertrr** [reply]
Yes, that's a useful flag, but some rules are disabled in the default config, so one still needs a way to discover them and to enable in the config file.

**Reply by cortinico** [reply]
True. I'd say `buildUponDefaultConfig` makes your life easier when maintaining your config file but loses in discoverability.

**Reply by BraisGabin** [reply]
I don't use `buildUponDefaultConfig` exactly for that reason. In 2.0 we need to redefine the configuration to make it easier to discover new rules and their configurations.

What do you think about enable all the rules by default so you need to opt out. And then some kind of autocompletation for the configurations. A kotlin dsl or a yaml schema seems like a good idea here.

**Reply by petertrr** [reply]
FWIW, in [diktat](https://github.com/cqfn/diktat) - our custom ruleset for ktlint - we are enabling all rules by default. Then you can explicitly disable or add configuration options for particular rules in yaml config. So when after update one starts seeing unwanted errors, they can disable them.
